File: backend/module9.py
from flask import Blueprint, jsonify, request
from flask_cors import CORS
import pandas as pd
import os
from datetime import datetime
from pymongo import MongoClient
import numpy as np

module9_bp = Blueprint('module9', __name__)
CORS(module9_bp)  # Enable CORS for this blueprint

# MongoDB client setup
client = MongoClient("mongodb://localhost:27017/")
db = client["data_cleaning_db"]
module9_collection = db["module9"]

UPLOAD_FOLDER = "uploads"

if not os.path.exists(UPLOAD_FOLDER):
    os.makedirs(UPLOAD_FOLDER)

def log_action(file_id, action, details):
    module9_collection.update_one(
        {"file_id": file_id},
        {"$push": {"actions_log": {"action": action, "details": details, "timestamp": datetime.now()}}}
    )

def update_progress(file_id, status):
    module9_collection.update_one(
        {"file_id": file_id},
        {"$set": {"status": status, "last_updated": datetime.now()}}
    )

@module9_bp.route('/api/module9/import', methods=['POST'])
def import_data():
    file = request.files.get('file')
    if not file:
        return jsonify({"status": "error", "message": "No file provided."}), 400

    file_path = os.path.join(UPLOAD_FOLDER, file.filename)
    file.save(file_path)

    try:
        data = pd.read_csv(file_path)
        file_id = str(datetime.timestamp(datetime.now()))

        module9_collection.insert_one({
            "file_id": file_id,
            "file_path": file_path,
            "imported_at": datetime.now(),
            "status": "Imported",
            "actions_log": []
        })

        update_progress(file_id, "Imported")
        log_action(file_id, "Import", {"file_name": file.filename})

        return jsonify({"status": "success", "file_id": file_id}), 200

    except Exception as e:
        return jsonify({"status": "error", "message": str(e)}), 500




# Example route for module 9
@module9_bp.route('/example', methods=['GET'])
# Example route for module 9
@module9_bp.route('/example', methods=['GET'])
def example():
    return jsonify({"message": "Module 9 works!"}), 200


@module9_bp.route('/api/module9/remove_duplicates', methods=['POST'])
def remove_duplicates():
    file_id = request.json.get('file_id')
    if not file_id:
        return jsonify({"status": "error", "message": "File ID is required."}), 400

    file_record = module9_collection.find_one({"file_id": file_id})
    if not file_record:
        return jsonify({"status": "error", "message": "File not found."}), 404

    try:
        data = pd.read_csv(file_record['file_path'])
        before_count = len(data)
        data = data.drop_duplicates()
        after_count = len(data)
        duplicate_count = before_count - after_count

        data.to_csv(file_record['file_path'], index=False)

        update_progress(file_id, "Duplicates Removed")
        log_action(file_id, "Remove Duplicates", {"duplicates_removed": duplicate_count})

        return jsonify({"status": "success", "duplicates_removed": duplicate_count}), 200

    except Exception as e:
        return jsonify({"status": "error", "message": str(e)}), 500

@module9_bp.route('/api/module9/fill_missing', methods=['POST'])
def fill_missing():
    file_id = request.json.get('file_id')
    strategy = request.json.get('strategy', 'mean')  # Default to 'mean'

    if not file_id:
        return jsonify({"status": "error", "message": "File ID is required."}), 400

    file_record = module9_collection.find_one({"file_id": file_id})
    if not file_record:
        return jsonify({"status": "error", "message": "File not found."}), 404

    try:
        # Load the dataset
        data = pd.read_csv(file_record['file_path'])

        # Handle missing data based on the strategy
        if strategy == 'mean':
            data.fillna(data.mean(), inplace=True)
        elif strategy == 'median':
            data.fillna(data.median(), inplace=True)
        elif strategy == 'mode':
            data.fillna(data.mode().iloc[0], inplace=True)
        else:
            return jsonify({"status": "error", "message": "Invalid strategy."}), 400

        # Save the updated data
        data.to_csv(file_record['file_path'], index=False)

        # Log the action
        update_progress(file_id, "Missing Data Filled")
        log_action(file_id, "Fill Missing", {"strategy": strategy})

        return jsonify({"status": "success", "message": f"Missing data filled using {strategy} strategy."}), 200

    except Exception as e:
        return jsonify({"status": "error", "message": str(e)}), 500


@module9_bp.route('/api/module9/normalize', methods=['POST'])
def normalize_data():
    file_id = request.json.get('file_id')

    if not file_id:
        return jsonify({"status": "error", "message": "File ID is required."}), 400

    file_record = module9_collection.find_one({"file_id": file_id})
    if not file_record:
        return jsonify({"status": "error", "message": "File not found."}), 404

    try:
        data = pd.read_csv(file_record['file_path'])
        numeric_columns = data.select_dtypes(include=['number']).columns
        data[numeric_columns] = (data[numeric_columns] - data[numeric_columns].mean()) / data[numeric_columns].std()

        data.to_csv(file_record['file_path'], index=False)

        update_progress(file_id, "Data Normalized")
        log_action(file_id, "Normalize Data", {"normalized_columns": list(numeric_columns)})

        return jsonify({"status": "success", "normalized_columns": list(numeric_columns)}), 200

    except Exception as e:
        return jsonify({"status": "error", "message": str(e)}), 500

@module9_bp.route('/api/module9/detect_outliers', methods=['POST'])
def detect_outliers():
    file_id = request.json.get('file_id')
    if not file_id:
        return jsonify({"status": "error", "message": "File ID is required."}), 400

    file_record = module9_collection.find_one({"file_id": file_id})
    if not file_record:
        return jsonify({"status": "error", "message": "File not found."}), 404

    try:
        data = pd.read_csv(file_record['file_path'])
        numeric_columns = data.select_dtypes(include=['number']).columns
        outlier_info = {}

        for col in numeric_columns:
            z_scores = np.abs((data[col] - data[col].mean()) / data[col].std())
            outliers = data[z_scores > 3]
            outlier_info[col] = len(outliers)

        log_action(file_id, "Detect Outliers", {"outliers_detected": outlier_info})
        update_progress(file_id, "Outliers Detected")

        return jsonify({"status": "success", "outliers_detected": outlier_info}), 200

    except Exception as e:
        return jsonify({"status": "error", "message": str(e)}), 500

@module9_bp.route('/api/module9/progress', methods=['GET'])
def track_cleaning_progress():
    file_id = request.args.get('file_id')
    if not file_id:
        return jsonify({"status": "error", "message": "File ID is required."}), 400

    progress = module9_collection.find_one({"file_id": file_id})
    if not progress:
        return jsonify({"status": "error", "message": "No progress found for this file."}), 404

    return jsonify({"status": "success", "progress": progress}), 200

File: frontend/src/module9.js
// src/components/Module9.js
import React, { useState } from 'react';
import axios from 'axios';

const Module9 = () => {
  const [fileId, setFileId] = useState('');
  const [status, setStatus] = useState('');
  const [message, setMessage] = useState('');
  
  const handleFileIdChange = (e) => setFileId(e.target.value);

  // Function to trigger the file import
  const importData = async (e) => {
    e.preventDefault();
    try {
      const formData = new FormData();
      formData.append('file', e.target.files[0]);
      
      const response = await axios.post('http://localhost:5000/api/module9/import', formData, {
        headers: { 'Content-Type': 'multipart/form-data' },
      });
      
      setStatus(response.data.status);
      setMessage(response.data.file_id);
    } catch (error) {
      setStatus('error');
      setMessage(error.message);
    }
  };

  const removeDuplicates = async () => {
    try {
      const response = await axios.post('http://localhost:5000/api/module9/remove_duplicates', { file_id: fileId });
      setStatus(response.data.status);
      setMessage(`Duplicates removed: ${response.data.duplicates_removed}`);
    } catch (error) {
      setStatus('error');
      setMessage(error.message);
    }
  };

  const fillMissingData = async (strategy) => {
    try {
      const response = await axios.post('http://localhost:5000/api/module9/fill_missing', {
        file_id: fileId,
        strategy,
      });
      setStatus(response.data.status);
      setMessage(response.data.message);
    } catch (error) {
      setStatus('error');
      setMessage(error.message);
    }
  };

  const normalizeData = async () => {
    try {
      const response = await axios.post('http://localhost:5000/api/module9/normalize', { file_id: fileId });
      setStatus(response.data.status);
      setMessage(`Normalized columns: ${response.data.normalized_columns.join(', ')}`);
    } catch (error) {
      setStatus('error');
      setMessage(error.message);
    }
  };

  const detectOutliers = async () => {
    try {
      const response = await axios.post('http://localhost:5000/api/module9/detect_outliers', { file_id: fileId });
      setStatus(response.data.status);
      setMessage(`Outliers detected: ${JSON.stringify(response.data.outliers_detected)}`);
    } catch (error) {
      setStatus('error');
      setMessage(error.message);
    }
  };

  const trackProgress = async () => {
    try {
      const response = await axios.get(`http://localhost:5000/api/module9/progress?file_id=${fileId}`);
      setStatus(response.data.status);
      setMessage(`Progress: ${JSON.stringify(response.data.progress)}`);
    } catch (error) {
      setStatus('error');
      setMessage(error.message);
    }
  };

  return (
    <div className="container mt-5">
      <h2 className="text-center">Module 9: Data Cleaning Operations</h2>
      <form>
        <div className="mb-3">
          <label htmlFor="fileId" className="form-label">File ID</label>
          <input
            type="text"
            className="form-control"
            id="fileId"
            value={fileId}
            onChange={handleFileIdChange}
            placeholder="Enter File ID"
          />
        </div>
        <div className="mb-3">
          <label htmlFor="fileInput" className="form-label">Import File</label>
          <input
            type="file"
            className="form-control"
            id="fileInput"
            onChange={importData}
          />
        </div>

        {/* Action Buttons */}
        <div className="d-flex justify-content-between">
          <button type="button" className="btn btn-primary" onClick={removeDuplicates}>
            Remove Duplicates
          </button>
          <button type="button" className="btn btn-warning" onClick={() => fillMissingData('mean')}>
            Fill Missing Data (Mean)
          </button>
          <button type="button" className="btn btn-info" onClick={normalizeData}>
            Normalize Data
          </button>
          <button type="button" className="btn btn-danger" onClick={detectOutliers}>
            Detect Outliers
          </button>
          <button type="button" className="btn btn-secondary" onClick={trackProgress}>
            Track Progress
          </button>
        </div>

        {/* Status and Message */}
        {status && (
          <div className={`alert alert-${status === 'error' ? 'danger' : 'success'} mt-4`}>
            <strong>{status === 'error' ? 'Error:' : 'Success:'}</strong> {message}
          </div>
        )}
      </form>
    </div>
  );
};

export default Module9;
